service/history/historyEngine.go

```diff
@@ -402,10 +424,24 @@ func (e *historyEngineImpl) registerDomainFailoverCallback() {
 			shardNotificationVersion := e.shard.GetDomainNotificationVersion()
 			failoverDomainIDs := map[string]struct{}{}
 
-			for _, nextDomain := range nextDomains {
+			for idx, nextDomain := range nextDomains {
 				failoverPredicate(shardNotificationVersion, nextDomain, func() {
 					failoverDomainIDs[nextDomain.GetInfo().ID] = struct{}{}
 				})
+
+				// handle graceful failover on active to passive
+				if err := failoverMarkerInsertion(
+					shardNotificationVersion,
+					prevDomains[idx],
+					nextDomain,
+				); err != nil {
+					e.logger.Error("Failed to insert failover marker to replication queue.",
+						tag.Error(err),
+						tag.WorkflowDomainName(nextDomain.GetInfo().Name))
+					e.metricsClient.IncCounter(metrics.HistoryFailoverMarkerScope, metrics.HistroyFailoverMarkerInsertFailure)
+					// fail this failover callback and it retries on next domain cache refresh
```

Comment:
you may need to change the domain cache callback function to return err so domain cache can revert the changes and redo this callback.

one more thing, the hand over seems to relying on edge triggering, maybe it is necessary to have the change history of the domain?

Reply:
Return error is one approach. I updated the position to insert this marker. So it will insert the marker after it calls the failover domain function in task processor. In this way, the domain can be updated even if the insert operation fails.